### PR TITLE
feat: added GitLab secret detection report parser

### PR DESCRIPTION
**Pull request by natebwangsut**
Added a feature to accept GitLab secret detection reports

Ref: https://docs.gitlab.com/ee/user/application_security/secret_detection/

### CODE REVIEW

**Comment by Maffooch** [comment]
Hi @natebwangsut this is a great start! Please follow our [how to guide ](https://github.com/DefectDojo/django-DefectDojo/blob/dev/docs/content/en/contributing/how-to-write-a-parser.md) to make it perfect. There has been a lot of changes around creating parsers recently!

File: dojo/fixtures/test_type.json [hunk 1]
```diff
@@ -726,5 +726,12 @@
         },
         "model": "dojo.test_type",
         "pk": 195
+    },
+    {
+        "fields": {
+            "name": "GitLab Secret Detection Report"
+        },
+        "model": "dojo.test_type",
+        "pk": 196
```
Comment (Maffooch) [inline dojo/fixtures/test_type.json:735]: Not needed any longer

**Review by Maffooch — CHANGES_REQUESTED** [review]

File: dojo/fixtures/test_type.json (hunk 1 above)
Comment (natebwangsut) [reply dojo/fixtures/test_type.json]: fixed

File: dojo/tools/gitlab_secret_detection_report/parser.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                static_finding=True,
+                dynamic_finding=False,
+            )
+
+            if file_path != "":
+                finding.file_path = file_path
+            if line != -1:
+                finding.line = line
+            if raw_source_code != "":
+                finding.references = raw_source_code
```
Comment (damiencarol) [inline dojo/tools/gitlab_secret_detection_report/parser.py:71]: `references` attribute is not a piece of code. It's more like a list of URL and/or Ids

File: dojo/tools/gitlab_secret_detection_report/parser.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+        return list(dupes.values())
+
+    def normalise_severity(self, severity):
+        """
+        Normalise GitLab's severity to DefectDojo's
+        (Critical, High, Medium, Low, Unknown, Info) -> (Critical, High, Medium, Low, Informational)
+        """
+        if severity == "Unknown" or severity == "Info":
+            return "Informational"
```
Comment (damiencarol) [inline dojo/tools/gitlab_secret_detection_report/parser.py:94]: `Informational` is not a good value for severity.
Check model.py for more info

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: dojo/tools/gitlab_secret_detection_report/parser.py (hunk 3 above)
Comment (natebwangsut) [reply dojo/tools/gitlab_secret_detection_report/parser.py]: fixed, kindly check the PR

File: dojo/tools/gitlab_secret_detection_report/parser.py (hunk 2 above)
Comment (natebwangsut) [reply dojo/tools/gitlab_secret_detection_report/parser.py]: fixed, kindly check the PR

File: dojo/tools/gitlab_secret_detection_report/parser.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+            finding = Finding(
+                test=test,
+                title=title,
+                description=description,
+                date=date,
+                severity=severity,
+                static_finding=True,
+                dynamic_finding=False,
+            )
```
Comment (damiencarol) [inline dojo/tools/gitlab_secret_detection_report/parser.py:64]: it seems that you have an id in the file. Could you set `unique_id_from_tool` ?

Comment (natebwangsut) [reply dojo/tools/gitlab_secret_detection_report/parser.py:66]: added

**Comment by natebwangsut** [comment]
@damiencarol let me know what do you think

File: dojo/tools/gitlab_secret_detection_report/parser.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                str(description + title).encode("utf-8")
+            ).hexdigest()
+            if dupe_key in dupes:
+                find = dupes[dupe_key]
+                if finding.description:
+                    find.description += "\n" + finding.description
+                find.unsaved_endpoints.extend(finding.unsaved_endpoints)
+                dupes[dupe_key] = find
+            else:
+                dupes[dupe_key] = finding
```
Comment (damiencarol) [inline dojo/tools/gitlab_secret_detection_report/parser.py:85]: As we are using a `unique_id_from_tool`, there should not have duplicate in the same report, no?
Also internal de-duplication should not be required.

File: dojo/unittests/tools/test_gitlab_secret_detection_report_parser.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        for finding in findings:
+            for endpoint in finding.unsaved_endpoints:
+                endpoint.clean()
+        first_finding = findings[0]
+        self.assertEqual(1, len(findings))
+        self.assertEqual(5, first_finding.line)
+        self.assertEqual("Critical", first_finding.severity)
+        self.assertEqual("README.md", first_finding.file_path)
+        self.assertEqual("AWS\nAKIAIOSFODNN7EXAMPLE", first_finding.description)
+
```
Comment (damiencarol) [inline dojo/unittests/tools/test_gitlab_secret_detection_report_parser.py:36]: a check on `unique_id_from_tool` is missing

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: dojo/tools/gitlab_secret_detection_report/parser.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import hashlib
```
Comment (damiencarol) [inline dojo/tools/gitlab_secret_detection_report/parser.py]:
```suggestion
```

**Review by damiencarol — APPROVED** [review]
Good work!

**Comment by damiencarol** [comment]
@Maffooch could you take another look?

**Comment by natebwangsut** [comment]
@damiencarol @Maffooch any update ?

File: dojo/tools/gitlab_secret_detection_report/parser.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        return "GitLab Secret Detection Report file can be imported in JSON format (option --json)."
+
+    def get_findings(self, file, test):
+        # Load JSON data from uploaded file
+        data = json.load(file)
+
+        findings = []
+
+        # This is required by schema - it won't be null / undefined
+        date = data["scan"]["end_time"]
```
Comment (StefanFl) [inline dojo/tools/gitlab_secret_detection_report/parser.py:27]: @natebwangsut In #4808 you implemented a bugfix for a similar line of code. Do we need the same here?

File: dojo/unittests/tools/test_gitlab_secret_detection_report_parser.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        for finding in findings:
+            for endpoint in finding.unsaved_endpoints:
+                endpoint.clean()
+        first_finding = findings[0]
+        self.assertEqual(1, len(findings))
+        self.assertEqual(5, first_finding.line)
+        self.assertEqual("Critical", first_finding.severity)
+        self.assertEqual("README.md", first_finding.file_path)
+        self.assertEqual("AWS\nAKIAIOSFODNN7EXAMPLE", first_finding.description)
+        self.assertEqual("714ed3e4e289ad35a089e0a888e8d0120b6a6083b1090a189cbc6a3227396240", first_finding.unique_id_from_tool)
```
Comment (StefanFl) [inline dojo/unittests/tools/test_gitlab_secret_detection_report_parser.py:36]: Please check the date as well.

**Review by StefanFl — CHANGES_REQUESTED** [review]

File: dojo/tools/gitlab_secret_detection_report/parser.py (hunk 8 above)
Comment (natebwangsut) [reply dojo/tools/gitlab_secret_detection_report/parser.py]: @StefanFl yup, i've also fixed them here as well

**Review by StefanFl — APPROVED** [review]

**Comment by damiencarol** [comment]
@Maffooch do you have more things to say for this PR? The contributor fixed your comment.

**Review by Maffooch — APPROVED** [review]

**Pull request merged**